### ISSUE TITLE
VPC endpoints module fails if a dynamodb block is added

### ISSUE THREAD

**Issue by ejimz** [terraform-aws-modules/terraform-aws-vpc#650]
## Description

if we use the kms block with another block (like dynamodb) in the vpc_endpoints module it returned an index error

```
Error: Invalid index
```

## Versions
 - Terraform: v0.13.7
 - VPC Endpoints version 3.0.0
 - Module:
       -provider[registry.terraform.io/hashicorp/aws] >= 3.15.*

## Reproduction
1. Add two or more vpc endpoints
2. One of these endpoints must be "kms"
3. Add a security_group_ids to the kms block
3. run "terraform plan"

## Actual behavior
After a Terraform plan we get the next error
```
Error: Invalid index

  on terraform-aws-vpc/modules/vpc-endpoints/main.tf line 25, in resource "aws_vpc_endpoint" "this":
  25:   service_name      = data.aws_vpc_endpoint_service.this[each.key].service_name
    |----------------
    | data.aws_vpc_endpoint_service.this is object with 1 attribute "dynamodb"
    | each.key is "kms"

The given key does not identify an element in this collection value.

Releasing state lock. This may take a few moments...
[terragrunt] 2021/05/31 15:35:18 Hit multiple errors:
exit status 1

```

## Additional context
Terraform code
```
module "vpc_endpoints" {
  source    = "terraform-aws-modules/vpc/aws//modules/vpc-endpoints"

  vpc_id    = module.vpc.vpc_id
  endpoints = {
    dynamodb = {
        service         = "dynamodb"
        service_type    = "Gateway"
        route_table_ids = flatten([module.vpc.private_route_table_ids])
      },
    kms = {
        service             = "kms"
        subnet_ids          = module.vpc.private_subnets
        security_group_ids  = [module.kms_https_sg.security_group_id]
      }
  }

  tags = var.tags
}

```

if we remove the "security_group_ids" line or use the kms block alone (without dynamodb block) it works

**Comment by ejimz** [comment]
the issue was the "security_group_ids" parameter in the kms block, after remove it started working

**Comment by ejimz** [comment]
If I use a global " security_group_ids" instead of a specific one in the kms block it works well

**Comment by mkielar** [comment]
Let me say that it's even worse with terraform 0.14.x. There are additional type-checks in 0.14 and up, which cause a flare of errors on `route_table_ids`, `security_group_ids` and `tags` attributes if they're missing, but multiple entries are added in `endpoints` map.

We're currently migrating 2x to 3x having a lot of dynamic, variable-driven conditional conditional logic around VPC Endpoints, and we're going to loop over vpc endpoins in a local map, and put them to the module one by one like below, because this thing is unusable otherwise...

```terraform
module "vpc_endpoints" {
  source  = "terraform-aws-modules/vpc/aws//modules/vpc-endpoints"
  version = "3.1.0"

  for_each = local.enabled_endpoints

  vpc_id             = module.vpc.vpc_id
  security_group_ids = [ try(aws_security_group.vpc_endpoint_sg[0].id, null) ]
  tags = merge(var.tags, {
    "Name" = var.vpc_name
  })

  endpoints = {
    endpoint : each.value
  }
}
```

Then we're consuming it with:
```
module.vpc_endpoints["s3"].endpoints["endpoint"].id
```

Ugly as hell, but seems to be working.

**Comment by antonbabenko** [comment]
Hmm, there are similar issues in other modules where attributes of **exact** type should be specified even if they are not gonna be used.

https://github.com/terraform-aws-modules/terraform-aws-route53/pull/39#issuecomment-859899157

**Comment by jmcorallo** [comment]
As a workaround you can create 2 instances of the module, one for Interface endpoints and another for Gateway endpoints, eg:
```
module "endpoints_gateway" {
  source  = "terraform-aws-modules/vpc/aws//modules/vpc-endpoints"
  version = "3.1.0"

  vpc_id = module.vpc.vpc_id

  endpoints = {
    s3 = {
      service      = "s3",
      service_type = "Gateway"
    },
    dynamodb = {
      service      = "dynamodb",
      service_type = "Gateway"
    },
  }
}

module "endpoints_interface" {
  source  = "terraform-aws-modules/vpc/aws//modules/vpc-endpoints"
  version = "3.1.0"

  vpc_id             = module.vpc.vpc_id
  security_group_ids = var.sg_ids
  subnet_ids         = module.vpc.private_subnets

  endpoints = {
    kms = {
      service             = "kms"
      private_dns_enabled = true
    },
    # others...
  }
}
```

**Comment by rogersd** [comment]
Yeah, this just keeps getting worse in 1.x:

```
╷
│ Error: Inconsistent conditional result types
│ 
│   on .terraform/modules/eks_vpc.vpc_endpoints/modules/vpc-endpoints/main.tf line 2, in locals:
│    2:   endpoints = var.create ? var.endpoints : tomap({})
│     ├────────────────
│     │ var.endpoints is object with 10 attributes
│ 
│ The true result value has the wrong type: element types must all match for
│ conversion to map.
╵
Operation failed: failed running terraform plan (exit 1)
```
Super vague and leads to a lot of tail chasing while you try to figure out what the error means.

**Comment by idelkysq** [comment]
Hi, I might have same issue. 

When I add a new block to create servicecatalog endpoint in an existing VPC, I'm  getting the following error:

I'm using Terraform v0.13.4

```
Error: Invalid index

  on .terraform/modules/vpc_endpoint/nested/vpc_endpoint_nested/main.tf line 21, in resource "aws_vpc_endpoint" "this":
  21:   service_name      = data.aws_vpc_endpoint_service.this[each.key].service_name
    |----------------
    | data.aws_vpc_endpoint_service.this is object with 9 attributes
    | each.key is "service_catalog"

The given key does not identify an element in this collection value.

ERRO[0075] 1 error occurred:
	* exit status 1
```

We are using the following tf module:

```
################################################################################
# Endpoint(s)
################################################################################

data "aws_vpc_endpoint_service" "this" {
  for_each = var.endpoints

  service      = lookup(each.value, "service", null)
  service_name = lookup(each.value, "service_name", null)

  filter {
    name   = "service-type"
    values = [lookup(each.value, "service_type", "Interface")]
  }
}

resource "aws_vpc_endpoint" "this" {
  for_each = var.endpoints

  vpc_id                      = var.vpc_id
  service_name          = data.aws_vpc_endpoint_service.this[each.key].service_name
  vpc_endpoint_type = lookup(each.value, "service_type", "Interface")
  auto_accept            = lookup(each.value, "auto_accept", null)

  security_group_ids  = lookup(each.value, "service_type", "Interface") == "Interface" ? distinct(concat(var.security_group_ids, lookup(each.value, "security_group_ids", []))) : null
  subnet_ids          = lookup(each.value, "service_type", "Interface") == "Interface" ? distinct(concat(var.subnet_ids, lookup(each.value, "subnet_ids", []))) : null
  route_table_ids     = lookup(each.value, "service_type", "Interface") == "Gateway" ? lookup(each.value, "route_table_ids", null) : null
  policy              = lookup(each.value, "policy", null)
  private_dns_enabled = lookup(each.value, "service_type", "Interface") == "Interface" ? lookup(each.value, "private_dns_enabled", null) : null

  tags = merge(var.tags, lookup(each.value, "tags", {}))

  timeouts {
    create = lookup(var.timeouts, "create", "10m")
    update = lookup(var.timeouts, "update", "10m")
    delete = lookup(var.timeouts, "delete", "10m")
  }
}
```
I'm not a Terraform expert, trying to get there.

What I'm missing in this case? Ty

**Comment by antonbabenko** [comment]
This issue has been resolved in version 3.11.2 :tada:

**Comment by rogersd** [comment]
Fantastic, thanks!
